[Dynamo][Guards]Add e2e user stack to guard debug info and ++ leaf guard

fxdawnn opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169999

Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

fxdawnn committed:
[Dynamo][Guards]Add user stack to guard debug info and leaf guard

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 reviewed (changes requested):
The direction is good. But this is vastly incomplete.


We want the plumb the python `guard.user_stack` into the LeafGuard. For example, in NONE_MATCH guard, you would change these lines

```
        self.get_guard_manager(guard).add_none_match_guard(
            get_verbose_code_parts(code, guard)
        )
```

```
        self.get_guard_manager(guard).add_none_match_guard(
            get_verbose_code_parts(code, guard), guard.user_stack
        )
```

so you will have to make many changes in both python and C++ files to plumb the user stack.

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -1663,6 +1687,7 @@ class LeafGuard {
   // This is set while constructing the leaf guard. This is used for identifying
   // the cause of recompilation.
   py::list _verbose_code_parts;
+  py::object* _user_stack;
```
It should not be a pointer.

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -7740,32 +7952,35 @@ PyObject* torch_c_dynamo_guards_init() {
           "add_dict_version_guard",
           [](DictGuardManager& self,
              py::object value,
              py::object verbose_code_parts) -> void {
```
no user stack?

fxdawnn replied:
I found two ``add_dict_version_guard`` def here. Should we remove this one?

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 commented on torch/_C/_dynamo/guards.pyi:
```diff
@@ -22,6 +23,7 @@ class GuardDebugInfo:
     verbose_code_parts: list[str]
     result: bool
     num_guards_executed: int
+    user_stack: Any
```
For now, you can say `Optional[Union[traceback.StackSummary, str]]`

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add user stack to guard debug info and leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add e2e user stack to guard debug info and ++ leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 reviewed (approved):
I have one comment, but this looks good. 

I am approving but please address the comment. Thanks for the PR.

fxdawnn committed:
Update on "[Dynamo][Guards]Add e2e user stack to guard debug info and ++ leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Add e2e user stack to guard debug info and ++ leaf guard"


Fix https://github.com/pytorch/pytorch/issues/167898

## Problem
Recompilation is expensive and it would be nice to reduce the number of such actions. The guard recompilations reason logs currently capture the last line of the stack and it doesn’t give enough information. By providing a full user stack trace, we would make finding the root cause of recompilation a lot easier and reduce the total amount of recompilations. 

## Solution
An additional field is being implemented in LeafGuard to store the complete user stack, which will then be extracted via Guard debug information. This approach is expected to be successful because it isolates specific guards and only provides the necessary information when a guard fails and recompiles.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
